test-data/unit/check-literal.test

```diff
@@ -2214,4 +2214,376 @@ UnicodeDict = TypedDict(b'UnicodeDict', {'key': int})
+reveal_type(f.instancevar3)          # E: Revealed type is 'Literal[True]'
+reveal_type(f.instancevar4)          # E: Revealed type is 'None'
+force1(reveal_type(f.instancevar1))  # E: Revealed type is 'Literal[1]'
+force2(reveal_type(f.instancevar2))  # E: Revealed type is 'Literal['foo']'
+force3(reveal_type(f.instancevar3))  # E: Revealed type is 'Literal[True]'
+force4(reveal_type(f.instancevar4))  # E: Revealed type is 'None'
+[builtins fixtures/primitives.pyi]
+[out]
+
+[case testLiteralFinalMismatchCausesError]
+from typing_extensions import Final, Literal
+
+var1: Final[Literal[4]]     = 1      # E: Incompatible types in assignment (expression has type "Literal[1]", variable has type "Literal[4]")
+var2: Final[Literal['bad']] = "foo"  # E: Incompatible types in assignment (expression has type "Literal['foo']", variable has type "Literal['bad']")
+var3: Final[Literal[False]] = True   # E: Incompatible types in assignment (expression has type "Literal[True]", variable has type "Literal[False]")
+
+class Foo:
+    classvar1: Final[Literal[4]]     = 1      # E: Incompatible types in assignment (expression has type "Literal[1]", variable has type "Literal[4]")
+    classvar2: Final[Literal['bad']] = "foo"  # E: Incompatible types in assignment (expression has type "Literal['foo']", variable has type "Literal['bad']")
+    classvar3: Final[Literal[False]] = True   # E: Incompatible types in assignment (expression has type "Literal[True]", variable has type "Literal[False]")
+
+    def __init__(self) -> None:
+        self.instancevar1: Final[Literal[4]]     = 1      # E: Incompatible types in assignment (expression has type "Literal[1]", variable has type "Literal[4]")
+        self.instancevar2: Final[Literal['bad']] = "foo"  # E: Incompatible types in assignment (expression has type "Literal['foo']", variable has type "Literal['bad']")
+        self.instancevar3: Final[Literal[False]] = True    # E: Incompatible types in assignment (expression has type "Literal[True]", variable has type "Literal[False]")
+
+# TODO: Fix the order in which these error messages are shown to be more consistent.
+var1 = 10                # E: Incompatible types in assignment (expression has type "Literal[10]", variable has type "Literal[4]") \
+                         # E: Cannot assign to final name "var1"
+
+Foo.classvar1 = 10       # E: Cannot assign to final attribute "classvar1" \
+                         # E: Incompatible types in assignment (expression has type "Literal[10]", variable has type "Literal[4]")
+
+Foo().instancevar1 = 10  # E: Cannot assign to final attribute "instancevar1" \
+                         # E: Incompatible types in assignment (expression has type "Literal[10]", variable has type "Literal[4]")
+[builtins fixtures/primitives.pyi]
+[out]
+
+[case testLiteralFinalGoesOnlyOneLevelDown]
+from typing import Tuple
+from typing_extensions import Final, Literal
+
+a: Final = 1
+b: Final = (1, 2)
```

Comment:
In principle we can support this at some point, but not now. IIRC there is already an issue about this (for common fields in named tuples).